engine/src/main/java/org/apache/apex/engine/plugin/AbstractApexPluginDispatcher.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.apex.stram.api.plugin.PluginManager;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileContext;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.service.AbstractService;
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
+
+import com.google.common.collect.Lists;
+
+import com.datatorrent.stram.StramAppContext;
+import com.datatorrent.stram.StreamingContainerManager;
+import com.datatorrent.stram.plan.logical.LogicalPlan;
+
+/**
+ * A default implementation for ApexPluginDispatcher. It handler common tasks such as per handler
+ * registration. actual dispatching is left for classes extending from it.
+ */
+public abstract class AbstractApexPluginDispatcher extends AbstractService implements ApexPluginDispatcher
+{
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractApexPluginDispatcher.class);
+  protected final Collection<ApexPlugin> plugins = Lists.newArrayList();
+  protected final StramAppContext appContext;
+  protected final StreamingContainerManager dmgr;
+  private final PluginLocator locator;
+  protected transient Configuration launchConfig;
+  protected FileContext fileContext;
+  protected final Map<ApexPlugin, PluginInfo> pluginInfoMap = new HashMap<>();
+  protected PluginContext pluginContext;
+
+  public AbstractApexPluginDispatcher(PluginLocator locator, StramAppContext context, StreamingContainerManager dmgr)
+  {
+    super(AbstractApexPluginDispatcher.class.getName());
+    this.locator = locator;
+    this.appContext = context;
+    this.dmgr = dmgr;
+    LOG.info("Creating apex service ");
+  }
+
+  private Configuration readLaunchConfiguration() throws IOException
+  {
+    Path appPath = new Path(appContext.getApplicationPath());
+    Path  configFilePath = new Path(appPath, LogicalPlan.LAUNCH_CONFIG_FILE_NAME);
+    try {
+      LOG.info("Reading launch configuration file ");
+      URI uri = appPath.toUri();
+      Configuration config = new YarnConfiguration();
+      fileContext = uri.getScheme() == null ? FileContext.getFileContext(config) : FileContext.getFileContext(uri, config);
+      FSDataInputStream is = fileContext.open(configFilePath);
+      config.addResource(is);
+      LOG.info("Read launch configuration");
+      return config;
+    } catch (FileNotFoundException ex) {
+      LOG.warn("Configuration file not found {}", configFilePath);
+      return new Configuration();
+    }
+  }
+
+  @Override
+  protected void serviceInit(Configuration conf) throws Exception
+  {
+    super.serviceInit(conf);
+    this.launchConfig = readLaunchConfiguration();
+    pluginContext = new DefaultPluginContextImpl(appContext, dmgr, this.launchConfig);
+    if (locator != null) {
+      Collection<ApexPlugin> plugins = locator.discoverPlugins();
+      if (plugins != null) {
+        this.plugins.addAll(plugins);
+      }
+    }
+
+    super.serviceStart();
```

Comment:
Why do you need to call serviceStart explicitly. Wouldn't it be called automatically by the framework.